src/main/java/com/aws/greengrass/lifecyclemanager/KernelAlternatives.java

```diff
@@ -162,31 +169,69 @@ public boolean isLaunchDirSetup() {
         return Files.isSymbolicLink(getCurrentDir()) && validateLaunchDirSetup(getCurrentDir());
     }
 
+    protected boolean canRecoverMissingLaunchDirSetup()
+            throws IOException, URISyntaxException, PackageLoadingException {
+        /*
+        Try and relink launch dir with the following replacement criteria
+        1. check if current Nucleus execution package is valid
+        2. check if any available Nucleus packages in component store is valid
+        3. fail with DirectoryValidationException if above steps do not satisfy
+         */
+        Path currentNucleusExecutablePath = locateCurrentKernelUnpackDir();
+        if (Files.exists(currentNucleusExecutablePath.resolve(KERNEL_BIN_DIR)
+                .resolve(Platform.getInstance().loaderFilename()))) {
+            logger.atDebug().kv("path", currentNucleusExecutablePath)
+                    .log("Current Nucleus executable is valid, setting up launch dir");
+            relinkInitLaunchDir(currentNucleusExecutablePath, true);
+            return true;
+        }
+
+        List<Path> localNucleusExecutablePaths = componentManager.findValidLocalNucleusPackage();
+        if (!localNucleusExecutablePaths.isEmpty()) {
```

Comment:
aren't you calling isEmpty on null object?

Reply:
Updated. This will still be an empty check in case artifacts dir is empty (initial installation) but this code path should only hit if a customer purposefully deletes the loader script from the unarchived dir. Imo, this is low risk.